src/censored.jl

```diff
@@ -0,0 +1,409 @@
+    end
+end
+
+function logpdf(d::Censored, x::Real)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    return if lower !== missing && x == lower
+        result = logcdf(d0, x)
+        _eqnotmissing(x, upper) ? zero(result) : result
+    elseif _eqnotmissing(x, upper)
+        _logccdf_inc(d0, x)
+    else
+        result = logpdf(d0, x)
+        _in_open_interval(x, lower, upper) ? result : oftype(result, -Inf)
+    end
+
+end
+
+function loglikelihood(d::Censored, x::AbstractArray{<:Real, M}) where {M}
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    log_prob_lower = lower === missing ? 0 : logcdf(d0, lower)
+    log_prob_upper = upper === missing ? 0 : _logccdf_inc(d0, upper)
+
+    return sum(x) do xi
+        _in_open_interval(xi, lower, upper) && return logpdf(d0, xi)
+        _eqnotmissing(xi, lower) && return log_prob_lower
+        _eqnotmissing(xi, upper) && return log_prob_upper
+        T = float(Base.promote_eltype(log_prob_lower, log_prob_upper))
+        return T(-Inf)
+    end
+end
+
+function cdf(d::Censored, x::Real)
+    result = cdf(d.uncensored, x)
+    return if d.lower !== missing && x < d.lower
+        zero(result)
+    elseif d.upper === missing || x < d.upper
+        result
+    else
+        one(result)
+    end
+end
+
+function logcdf(d::Censored, x::Real)
+    result = logcdf(d.uncensored, x)
+    return if d.lower !== missing && x < d.lower
+        oftype(result, -Inf)
+    elseif d.upper === missing || x < d.upper
+        result
+    else
+        zero(result)
+    end
+end
+
+function ccdf(d::Censored, x::Real)
+    result = ccdf(d.uncensored, x)
+    return if d.lower !== missing && x < d.lower
+        one(result)
+    elseif d.upper === missing || x < d.upper
+        result
+    else
+        zero(result)
+    end
+end
+
+function logccdf(d::Censored, x::Real)
+    result = logccdf(d.uncensored, x)
+    return if d.lower !== missing && x < d.lower
+        zero(result)
+    elseif d.upper === missing || x < d.upper
+        result
+    else
+        oftype(result, -Inf)
+    end
+end
+
+
+#### Sampling
+
+rand(rng::AbstractRNG, d::Censored) = _clamp(rand(rng, d.uncensored), d.lower, d.upper)
+
+
+#### Utilities
+
+# utilities to handle intervals represented with possibly missing bounds
+
+@inline _is_non_empty_interval(l::Real, u::Real) = l ≤ u
+
+@inline _in_open_interval(x::Real, l::Real, u::Real) = l < x < u
+@inline _in_open_interval(x::Real, ::Missing, u::Real) = x < u
+@inline _in_open_interval(x::Real, l::Real, ::Missing) = x > l
+
+function _to_truncated(d::Censored)
+    return truncated(
+        d.uncensored,
+        d.lower === missing ? -Inf : d.lower,
+        d.upper === missing ? Inf : d.upper,
+    )
+end
+
+_clamp(x, l, u) = clamp(x, l, u)
+_clamp(x, ::Missing, u) = min(x, u)
+_clamp(x, l, ::Missing) = max(x, l)
+
+@inline function _eqnotmissing(x, y)
+    result = x == y
+    return result === missing ? false : result
+end
```

Comment:
It seems we only call it as `_eqnotmissing(x::Real, y::Union{Real,Missing})`, so it would be easier to define
```suggestion
@inline _eqnotmissing(x::Real, y::Real) = x == y
@inline _eqnotmissing(::Real, ::Missing) = false
```
(or without `@inline` :wink:)